File: src/main/java/com/camackenzie/exvi/core/api/APIRequest.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.api;

import com.camackenzie.exvi.core.util.FutureWrapper;
import com.google.gson.Gson;
import java.io.Reader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpRequest.Builder;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.util.HashMap;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 *
 * @author callum
 */
public class APIRequest<T> {

    private static final Gson gson = new Gson();

    private T body;
    private transient HashMap<String, String> headers;
    private transient String endpoint;

    public APIRequest(String endpoint, T body, HashMap<String, String> headers) {
        this.endpoint = endpoint;
        this.body = body;
        this.headers = headers;
    }

    public APIRequest(String endpoint, T body) {
        this(endpoint, body, new HashMap<>());
    }

    public APIRequest<T> withJsonHeader() {
        this.headers.put("content-type", "application/json");
        return this;
    }

    public T getBody() {
        return this.body;
    }

    public void setBody(T body) {
        this.body = body;
    }

    public HashMap<String, String> getHeaders() {
        return this.headers;
    }

    public void setHeaders(HashMap<String, String> headers) {
        this.headers = headers;
    }

    public String getEndpoint() {
        return this.endpoint;
    }

    public void setEndpoint(String e) {
        this.endpoint = e;
    }

    public <E> FutureWrapper<APIResult<E>> send(Class<E> resultClass) {
        HttpClient client = HttpClient.newHttpClient();
        Builder requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(this.endpoint));
        for (var header : this.headers.entrySet()) {
            requestBuilder.setHeader(header.getKey(),
                    header.getValue());
        }
        requestBuilder.POST(BodyPublishers.ofString(gson.toJson(this.body)));

        return new FutureWrapper(new APIResultFutureWrapper<E>(resultClass,
                client.sendAsync(requestBuilder.build(), BodyHandlers.ofString())));
    }

    public static <T> APIRequest<T> fromJson(String json) {
        return gson.fromJson(json, APIRequest.class);
    }

    public static <T> APIRequest<T> fromJson(Reader json) {
        return gson.fromJson(json, APIRequest.class);
    }

    public static <T, E> FutureWrapper<APIResult<E>> send(String endpoint,
            T body,
            HashMap<String, String> headers,
            Class<E> resultClass) {
        return new APIRequest(endpoint, body, headers).send(resultClass);
    }

    public static <T, E> FutureWrapper<APIResult<E>> sendJson(String endpoint, T body,
            Class<E> resultClass) {
        return new APIRequest(endpoint, body).withJsonHeader().send(resultClass);
    }

    private static class APIResultFutureWrapper<T> implements Future<APIResult<T>> {

        private final Future<HttpResponse<String>> requestFuture;
        private final Class<T> tClass;
        private APIResult<T> result = null;

        public APIResultFutureWrapper(Class<T> tClass,
                CompletableFuture<HttpResponse<String>> sendAsync) {
            this.requestFuture = sendAsync;
            this.tClass = tClass;
        }

        @Override
        public boolean cancel(boolean mayInterruptIfRunning) {
            return this.requestFuture.cancel(mayInterruptIfRunning);
        }

        @Override
        public boolean isCancelled() {
            return this.requestFuture.isCancelled();
        }

        @Override
        public boolean isDone() {
            return this.requestFuture.isDone();
        }

        @Override
        public APIResult<T> get() throws InterruptedException, ExecutionException {
            return this.getFromInternalFuture(this.requestFuture.get());
        }

        @Override
        public APIResult<T> get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
            return this.getFromInternalFuture(this.requestFuture.get(timeout, unit));
        }

        private APIResult<T> getFromInternalFuture(HttpResponse<String> resp) {
            System.out.println("Internal future returned " + resp.body());
            if (this.result == null) {
                String responseBody = resp.body();
                try {
                    // Format headers
                    HashMap<String, String> headers = new HashMap<>();
                    this.requestFuture.get()
                            .headers()
                            .map()
                            .entrySet()
                            .stream()
                            .forEach(en -> {
                                String headerVal = en.getValue().size() == 0 ? "" : en.getValue().get(0);
                                headers.put(en.getKey(), headerVal);
                            });
                    // Return proper api result
                    return new APIResult(this.requestFuture.get().statusCode(),
                            gson.fromJson(responseBody, this.tClass), headers);
                } catch (InterruptedException e) {
                    System.out.println(e);
                    Thread.currentThread().interrupt();
                } catch (ExecutionException e) {
                    System.err.println(e);
                }
                return null;
            } else {
                return this.result;
            }
        }

        public Class<T> getResultClass() {
            return this.tClass;
        }

    }

}

File: exvi-core/src/commonMain/kotlin/com/camackenzie/exvi/core/api/APIRequest.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.api

import com.camackenzie.exvi.core.model.ExviSerializer
import com.camackenzie.exvi.core.util.CryptographyUtils
import com.camackenzie.exvi.core.util.ExviLogger
import com.camackenzie.exvi.core.util.SelfSerializable
import com.camackenzie.exvi.core.util.cached
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.statement.HttpResponse
import io.ktor.utils.io.core.*
import kotlinx.coroutines.*
import kotlinx.serialization.Polymorphic
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlin.jvm.JvmOverloads
import kotlin.jvm.JvmStatic

/**
 *
 * @author callum
 */
@Serializable
@Suppress("unused")
class APIRequest<T : SelfSerializable> {
    @Polymorphic
    val body: T

    @kotlin.jvm.Transient
    @Transient
    var headers: HashMap<String, String> = HashMap()

    @kotlin.jvm.Transient
    @Transient
    var endpoint: String = ""

    @JvmOverloads
    constructor(endpoint: String, body: T, headers: HashMap<String, String> = HashMap()) {
        this.endpoint = endpoint
        this.body = body
        this.headers = headers
    }

    fun toJson() = ExviSerializer.toJson(this)

    @JvmOverloads
    constructor(body: T, headers: HashMap<String, String> = HashMap()) : this("", body, headers)

    constructor(other: APIRequest<*>, newBody: T) {
        endpoint = other.endpoint
        body = newBody
        headers = other.headers
    }

    fun withJsonHeader(): APIRequest<T> {
        headers["content-type"] = "application/json"
        return this
    }

    fun sendAsync(
        coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.Default),
        dispatcher: CoroutineDispatcher = Dispatchers.Default,
        callback: (APIResult<String>) -> Unit
    ): Job = coroutineScope.launch(dispatcher) {
        send(callback)
    }

    fun hasEndpoint(): Boolean = endpoint.isBlank()

    suspend fun send(callback: (APIResult<String>) -> Unit) =
        send { response, body ->
            val parsedResponse: APIResult<String> =
                APIResult(
                    response?.status?.value ?: 418,
                    APIResult.decodeBody(body),
                    HashMap()
                )
            callback(parsedResponse)
        }

    suspend fun send(callback: (HttpResponse?, String) -> Unit) =
        try {
            HttpClient {
                expectSuccess = false
            }.use { httpClient ->
                val reqHeaders = headers
                val reqBody = body.toJson().cached().getEncoded()
                val response = httpClient.post<HttpResponse>(endpoint) {
                    headers {
                        for ((key, value) in reqHeaders) {
                            append(key, value)
                        }
                    }
                    body = reqBody
                }
                callback(response, response.receive())
            }
        } catch (e: Exception) {
            ExviLogger.e(e, tag = "CORE") { "Request failed" }
            callback(null, "Request error.")
        }

    companion object {
        @JvmStatic
        suspend fun <T : SelfSerializable> request(
            endpoint: String,
            body: T,
            headers: HashMap<String, String> = jsonHeaders(),
            callback: (APIResult<String>) -> Unit
        ) = APIRequest(endpoint, body, headers).send(callback)

        @JvmStatic
        fun <T : SelfSerializable> requestAsync(
            endpoint: String,
            body: T,
            headers: HashMap<String, String> = jsonHeaders(),
            coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.Default),
            coroutineDispatcher: CoroutineDispatcher = Dispatchers.Default,
            callback: (APIResult<String>) -> Unit
        ): Job = APIRequest(endpoint, body, headers)
            .sendAsync(coroutineScope, coroutineDispatcher, callback)

        @JvmStatic
        fun jsonHeaders(): HashMap<String, String> {
            val ret: HashMap<String, String> = HashMap()
            ret["content-type"] = "application/json"
            return ret
        }
    }

}
